File: sdl test/Life.cpp
//
//  Life.cpp
//  sdl test
//
//  Created by Richard on 12/08/2018.
//  Copyright © 2018 Richard. All rights reserved.
//

#include "Life.hpp"


Life::Life(int w, int h)
{
    width = w; height = h;
    grid.resize(width, std::vector<bool>(height, false));
    updateGrid.resize(width, std::vector<bool>(height, false));
}


void Life::clear()
{
    grid.clear();
    grid.resize(width, std::vector<bool>(height, false));
}

bool Life::isAlive(int x, int y)
{
    x = (x + width) % width;
    y = (y + height) % height;
    
    return grid[x][y];
}

bool Life::calculateState(int x, int y)
{
    int alive = 0;
    
    for (int i = -1; i <= 1; i++)
    {
        for (int j = -1; j <= 1; j++)
        {
            if ((i || j) && isAlive(x+i, y+j))
            {
                alive++;
            }
        }
    }
    
    
    if (alive == 2) return grid[x][y];
    if (alive == 3) return true;
    if (alive < 2) return false;
    if (alive > 3) return false;
    
    return false;
}

void Life::setCell(int x, int y, bool val)
{
    grid[x][y] = val;
}

void Life::iterate()
{
    
    for(int y = 0; y < height; y++)
    {
        for(int x = 0; x < width; x++)
        {
            updateGrid[x][y] = calculateState(x, y);
        }
    }
    
    grid = updateGrid;
    updateGrid.resize(width, std::vector<bool>(height, false));

}

File: life.lua

name = "Game of Life"   --global name field (required)
version = 1           --global version field(required)
--probably more things when I think of them


colors = {              --global colour table (required)
    {0,0,0},            --0: dead, black
    {255,255,255},      --1: alive, white
}

width = 0
height = 0

function isAlive(grid, x, y)
    xx = ((x-1 + width) % width) + 1
    yy = ((y-1 + height) % height) + 1

    if grid[xx][yy] == 0 then return false else return true end

end

function calculateState(grid, x, y)

    alive = 0

    for i = -1, 1 do
        for j = -1, 1 do
            if (i ~= 0 or j ~= 0) and isAlive(grid, x+i, y+j) then 
              alive = alive + 1 
            end
        end
    end

    if alive == 2 then return grid[x][y] end
    if alive == 3 then return 1 end
    
    return 0

end

--iterate function required, called each iteration
--current state of the grid is passed as a parameter
function iterate(grid)

    width = #grid
    height = #grid[1]    

    local updateGrid = {}      --table to be returned
    --init table
    for x=1,width do
        updateGrid[x] = {}
      for y=1,height do
        updateGrid[x][y] = 0
      end
    end

    --for each cell
    for y = 1, height do
        for x = 1, width do
            updateGrid[x][y] = calculateState(grid, x, y)
        end
    end

    return updateGrid --iterate must return a new grid of the same size as the original
end
